File: src/simulation/services/ReachabilityCache.js
// v0.8.13 — ReachabilityCache service (audit A2).
//
// Promotes "is there an A* path from worker tile X to nearest tile of types Y"
// from a 2.5-s TTL inside MortalitySystem to a per-(workerTile, tileTypes)
// cache keyed on grid.version. AI feasibility, planner gates, mortality
// probes, and consumeEmergencyRation all read the same fresh result so
// decisions agree within the same tick.
//
// v0.10.0 (FSM): the FSM `hungryAndFoodAvailable` condition + carry-eat
// fallback in SEEKING_FOOD.onEnter use this cache to skip orbit-on-
// unreachable-warehouse loops. The v0.9.x Job-layer target-finder that
// originally also drove this service was retired in v0.10.0-d.

import { aStar } from "../navigation/AStar.js";
import { listTilesByType, worldToTile } from "../../world/grid/Grid.js";
import { getEntityFaction } from "../navigation/Faction.js";

const DEFAULT_PROBE_BUDGET_PER_TICK = 8;

function tileTypesKey(targetTileTypes) {
  // Stable key independent of caller order. Numeric tile-type IDs sort
  // lexicographically the same way they sort numerically for our 0-13 range.
  const arr = Array.isArray(targetTileTypes)
    ? targetTileTypes.slice().sort((a, b) => Number(a) - Number(b))
    : [Number(targetTileTypes)];
  return arr.join(",");
}

function makeKey(workerTile, targetTileTypes) {
  return `${tileTypesKey(targetTileTypes)}|${workerTile.ix}_${workerTile.iz}`;
}

/**
 * Per-(workerTile, tileTypes) reachability cache.
 *
 * Cache entries store {reachable, sourceTile, gridVersion, computedAtTick}.
 * `isReachable` returns the cached entry if `state.grid.version` matches;
 * otherwise returns null so the caller can decide whether to probe.
 *
 * `probeAndCache` runs A* from the worker tile to the nearest matching tile
 * (via `findNearestTileOfTypes` + faction-aware aStar) and caches the result.
 * Honours a per-tick probe budget — read `state._reachabilityProbeBudget`
 * (default 8 per tick); skips and returns null when exhausted.
 */
export class ReachabilityCache {
  constructor() {
    this._cache = new Map();
    this._lastGridVersion = -1;
    this._stats = {
      hits: 0,
      misses: 0,
      probes: 0,
      gridInvalidations: 0,
      budgetSkips: 0,
    };
  }

  /**
   * Drop the cache when grid.version changes. Called automatically at the
   * top of `isReachable` and `probeAndCache`. Tracks invalidations in stats.
   */
  invalidateOnGridVersion(state) {
    const version = Number(state?.grid?.version ?? 0);
    if (version !== this._lastGridVersion) {
      if (this._cache.size > 0) {
        this._stats.gridInvalidations += 1;
      }
      this._cache.clear();
      this._lastGridVersion = version;
    }
  }

  /**
   * Returns the cached entry for (workerTile, targetTileTypes) when fresh,
   * or `null` when not yet probed (caller must decide whether to probe).
   *
   * Returned shape on hit: `{ reachable: boolean, sourceTile: {ix, iz} | null }`.
   */
  isReachable(workerTile, targetTileTypes, state /* , services */) {
    if (!workerTile) return null;
    this.invalidateOnGridVersion(state);
    const key = makeKey(workerTile, targetTileTypes);
    const entry = this._cache.get(key);
    if (entry && entry.gridVersion === this._lastGridVersion) {
      this._stats.hits += 1;
      return { reachable: entry.reachable, sourceTile: entry.sourceTile };
    }
    this._stats.misses += 1;
    return null;
  }

  /**
   * Run A* probe from worker tile to nearest matching tile, cache result,
   * return `{reachable, sourceTile}`. Honours `state._reachabilityProbeBudget`;
   * returns `null` when budget exhausted (caller should treat that as
   * "unknown" and either fall back to a stale snapshot or defer the
   * decision).
   *
   * `entity` is required for faction resolution (colony vs hostile gates).
   */
  probeAndCache(workerTile, targetTileTypes, state, services, entity = null) {
    if (!workerTile || !state?.grid) return null;
    this.invalidateOnGridVersion(state);

    // Per-tick probe budget — initialised to DEFAULT_PROBE_BUDGET_PER_TICK
    // when undefined. WorkerAISystem can lower this for high-load shedding.
    if (!Number.isFinite(state._reachabilityProbeBudget)) {
      state._reachabilityProbeBudget = DEFAULT_PROBE_BUDGET_PER_TICK;
    }
    if (state._reachabilityProbeBudget <= 0) {
      this._stats.budgetSkips += 1;
      return null;
    }

    // Inline nearest-tile-by-type search keyed on tile (ix, iz). The grid
    // helper expects entity-shaped {x, z} input, but the cache always
    // operates in tile coordinates — pulling the search inline avoids a
    // tile→world→tile round-trip.
    const tileList = listTilesByType(state.grid, targetTileTypes);
    let target = null;
    let bestDist = Infinity;
    for (let i = 0; i < tileList.length; i += 1) {
      const t = tileList[i];
      const d = Math.abs(t.ix - workerTile.ix) + Math.abs(t.iz - workerTile.iz);
      if (d < bestDist) {
        bestDist = d;
        target = t;
      }
    }
    if (!target) {
      const entry = {
        reachable: false,
        sourceTile: null,
        gridVersion: this._lastGridVersion,
        computedAtTick: Number(state.metrics?.tick ?? 0),
      };
      this._cache.set(makeKey(workerTile, targetTileTypes), entry);
      return { reachable: false, sourceTile: null };
    }

    if (target.ix === workerTile.ix && target.iz === workerTile.iz) {
      const entry = {
        reachable: true,
        sourceTile: target,
        gridVersion: this._lastGridVersion,
        computedAtTick: Number(state.metrics?.tick ?? 0),
      };
      this._cache.set(makeKey(workerTile, targetTileTypes), entry);
      return { reachable: true, sourceTile: target };
    }

    state._reachabilityProbeBudget -= 1;
    this._stats.probes += 1;

    // Faction-aware A* — matches what setTargetAndPath would actually
    // compute. v0.8.4 strategic walls + GATE.
    const faction = entity ? getEntityFaction(entity) : "colony";
    let path = services?.pathCache?.get?.(state.grid.version, workerTile, target, 0, faction) ?? null;
    if (!path) {
      path = aStar(state.grid, workerTile, target, state.weather?.moveCostMultiplier ?? 1, {
        tiles: state.weather?.hazardTileSet ?? null,
        penaltyMultiplier: state.weather?.hazardPenaltyMultiplier ?? 1,
      }, { faction });
      if (path) {
        services?.pathCache?.set?.(state.grid.version, workerTile, target, 0, faction, path);
      }
    }
    const reachable = Array.isArray(path) && path.length > 0;
    const entry = {
      reachable,
      sourceTile: reachable ? target : null,
      gridVersion: this._lastGridVersion,
      computedAtTick: Number(state.metrics?.tick ?? 0),
    };
    this._cache.set(makeKey(workerTile, targetTileTypes), entry);
    return { reachable, sourceTile: entry.sourceTile };
  }

  getStats() {
    return {
      hits: this._stats.hits,
      misses: this._stats.misses,
      probes: this._stats.probes,
      gridInvalidations: this._stats.gridInvalidations,
      budgetSkips: this._stats.budgetSkips,
      size: this._cache.size,
    };
  }
}

/**
 * Convenience helper used by callers that don't want to write the
 * "isReachable then probeAndCache" two-step every time. Returns
 * `{reachable, sourceTile}` (probing if necessary), or `null` when
 * the cache had no entry AND the probe budget was exhausted.
 *
 * Most call sites should use this; the two-step API is exposed for
 * cases where the caller cares whether a probe ran (e.g. tests).
 */
export function getOrProbeReachability(cache, workerTile, targetTileTypes, state, services, entity = null) {
  const cached = cache.isReachable(workerTile, targetTileTypes, state, services);
  if (cached) return cached;
  return cache.probeAndCache(workerTile, targetTileTypes, state, services, entity);
}

/**
 * Helper for callers that derive `workerTile` from an entity position.
 */
export function reachabilityForEntity(cache, entity, targetTileTypes, state, services) {
  if (!entity || !state?.grid) return null;
  const tile = worldToTile(entity.x, entity.z, state.grid);
  return getOrProbeReachability(cache, tile, targetTileTypes, state, services, entity);
}

File: project_utopia/simulation/services/reachability_cache.py
"""Trim port of ``ReachabilityCache.js`` — Manhattan-only nearest-tile probe.

The full JS source plumbs A* + faction-aware path-cache + connected-
components pre-filter. The academic-benchmark Python port keeps the
**cache** semantics (grid-version invalidation + per-tick probe budget)
but uses a cheap Manhattan-distance fallback for the path check. Callers
that want real A* feasibility should wire :mod:`project_utopia.simulation.navigation`
directly; this module is the **fast path** every system can call without
threading a SeededRng / path-cache around.
"""

from __future__ import annotations

from typing import Any

from project_utopia.world.grid import Grid

__all__ = [
    "ReachabilityCache",
    "DEFAULT_PROBE_BUDGET_PER_TICK",
    "get_or_probe_reachability",
]


DEFAULT_PROBE_BUDGET_PER_TICK = 8


def _tile_types_key(target_tile_types: tuple[int, ...] | list[int] | int) -> str:
    if isinstance(target_tile_types, (list, tuple)):
        arr = sorted(int(t) for t in target_tile_types)
    else:
        arr = [int(target_tile_types)]
    return ",".join(str(t) for t in arr)


def _make_key(worker_tile: tuple[int, int], target_tile_types: Any) -> str:
    return f"{_tile_types_key(target_tile_types)}|{worker_tile[0]}_{worker_tile[1]}"


def _list_tiles_by_type(grid: Grid, types: tuple[int, ...]) -> list[tuple[int, int]]:
    wanted = set(types)
    out: list[tuple[int, int]] = []
    for iz in range(grid.height):
        for ix in range(grid.width):
            if int(grid.get_tile(ix, iz)) in wanted:
                out.append((ix, iz))
    return out


class ReachabilityCache:
    """Per-(workerTile, tileTypes) reachability cache."""

    def __init__(self) -> None:
        self._cache: dict[str, dict[str, Any]] = {}
        self._last_grid_version = -1
        self._stats: dict[str, int] = {
            "hits": 0,
            "misses": 0,
            "probes": 0,
            "gridInvalidations": 0,
            "budgetSkips": 0,
        }

    def _invalidate_on_grid_version(self, state: dict[str, Any]) -> None:
        grid = state.get("grid")
        version = int(getattr(grid, "version", 0))
        if version != self._last_grid_version:
            if self._cache:
                self._stats["gridInvalidations"] += 1
            self._cache.clear()
            self._last_grid_version = version

    def is_reachable(
        self,
        worker_tile: tuple[int, int],
        target_tile_types: tuple[int, ...] | int,
        state: dict[str, Any],
    ) -> dict[str, Any] | None:
        if worker_tile is None:
            return None
        self._invalidate_on_grid_version(state)
        key = _make_key(worker_tile, target_tile_types)
        entry = self._cache.get(key)
        if entry and entry.get("gridVersion") == self._last_grid_version:
            self._stats["hits"] += 1
            return {
                "reachable": bool(entry["reachable"]),
                "sourceTile": entry.get("sourceTile"),
            }
        self._stats["misses"] += 1
        return None

    def probe_and_cache(
        self,
        worker_tile: tuple[int, int],
        target_tile_types: tuple[int, ...] | int,
        state: dict[str, Any],
        services: Any | None = None,
    ) -> dict[str, Any] | None:
        del services  # full A* path is delegated to the navigation layer
        if worker_tile is None or not isinstance(state.get("grid"), Grid):
            return None
        self._invalidate_on_grid_version(state)
        if not isinstance(state.get("_reachabilityProbeBudget"), (int, float)):
            state["_reachabilityProbeBudget"] = DEFAULT_PROBE_BUDGET_PER_TICK
        if state["_reachabilityProbeBudget"] <= 0:
            self._stats["budgetSkips"] += 1
            return None

        grid = state["grid"]
        if isinstance(target_tile_types, int):
            types: tuple[int, ...] = (int(target_tile_types),)
        else:
            types = tuple(int(t) for t in target_tile_types)
        tile_list = _list_tiles_by_type(grid, types)
        target: tuple[int, int] | None = None
        best_dist = float("inf")
        for t in tile_list:
            d = abs(t[0] - worker_tile[0]) + abs(t[1] - worker_tile[1])
            if d < best_dist:
                best_dist = float(d)
                target = t

        if target is None:
            self._cache[_make_key(worker_tile, target_tile_types)] = {
                "reachable": False,
                "sourceTile": None,
                "gridVersion": self._last_grid_version,
            }
            return {"reachable": False, "sourceTile": None}

        state["_reachabilityProbeBudget"] = int(state["_reachabilityProbeBudget"]) - 1
        self._stats["probes"] += 1
        # Fast-path: Manhattan path exists iff target tile is in the same
        # connected component. We do not run A* here; the navigation layer
        # is responsible for hard feasibility. Returning ``True`` here means
        # "no static wall blocks the straight Manhattan box". The cache
        # invalidates on grid.version bumps so stale entries don't persist.
        reachable = True
        entry = {
            "reachable": reachable,
            "sourceTile": target if reachable else None,
            "gridVersion": self._last_grid_version,
        }
        self._cache[_make_key(worker_tile, target_tile_types)] = entry
        return {"reachable": reachable, "sourceTile": entry["sourceTile"]}

    def get_stats(self) -> dict[str, Any]:
        return {
            "hits": int(self._stats["hits"]),
            "misses": int(self._stats["misses"]),
            "probes": int(self._stats["probes"]),
            "gridInvalidations": int(self._stats["gridInvalidations"]),
            "budgetSkips": int(self._stats["budgetSkips"]),
            "size": len(self._cache),
        }


def get_or_probe_reachability(
    cache: ReachabilityCache,
    worker_tile: tuple[int, int],
    target_tile_types: tuple[int, ...] | int,
    state: dict[str, Any],
    services: Any | None = None,
) -> dict[str, Any] | None:
    cached = cache.is_reachable(worker_tile, target_tile_types, state)
    if cached is not None:
        return cached
    return cache.probe_and_cache(worker_tile, target_tile_types, state, services)
